api/address/routes.go

```diff
@@ -357,8 +372,104 @@ func GetESDTBalance(c *gin.Context) {
 
 	tokenData := esdtTokenData{
 		TokenIdentifier: tokenIdentifier,
-		Balance:         balance,
-		Properties:      freeze,
+		Balance:         esdtData.Value.String(),
+		Properties:      string(esdtData.Properties),
+	}
+
+	c.JSON(
+		http.StatusOK,
+		shared.GenericAPIResponse{
+			Data:  gin.H{"tokenData": tokenData},
+			Error: "",
+			Code:  shared.ReturnCodeSuccess,
+		},
+	)
+}
+
+// GetESDTNFTData return the nft data for the given token
+func GetESDTNFTData(c *gin.Context) {
+	facade, ok := getFacade(c)
+	if !ok {
+		return
+	}
+
+	addr := c.Param("address")
+	if addr == "" {
+		c.JSON(
+			http.StatusBadRequest,
+			shared.GenericAPIResponse{
+				Data:  nil,
+				Error: fmt.Sprintf("%s: %s", errors.ErrGetESDTNFTData.Error(), errors.ErrEmptyAddress.Error()),
+				Code:  shared.ReturnCodeRequestError,
+			},
+		)
+		return
+	}
+
+	tokenIdentifier := c.Param("tokenIdentifier")
+	if tokenIdentifier == "" {
+		c.JSON(
+			http.StatusBadRequest,
+			shared.GenericAPIResponse{
+				Data:  nil,
+				Error: fmt.Sprintf("%s: %s", errors.ErrGetESDTNFTData.Error(), errors.ErrEmptyKey.Error()),
+				Code:  shared.ReturnCodeRequestError,
+			},
+		)
+		return
+	}
+
+	nonceAsStr := c.Param("nonce")
+	if nonceAsStr == "" {
+		c.JSON(
+			http.StatusBadRequest,
+			shared.GenericAPIResponse{
+				Data:  nil,
+				Error: fmt.Sprintf("%s: %s", errors.ErrNonceInvalid.Error(), errors.ErrEmptyKey.Error()),
+				Code:  shared.ReturnCodeRequestError,
+			},
+		)
+		return
+	}
+
+	nonceAsBigInt, okConvert := big.NewInt(0).SetString(nonceAsStr, 10)
+	if !okConvert {
+		c.JSON(
+			http.StatusBadRequest,
+			shared.GenericAPIResponse{
+				Data:  nil,
+				Error: fmt.Sprintf("%s: %s", errors.ErrNonceInvalid.Error(), errors.ErrEmptyKey.Error()),
+				Code:  shared.ReturnCodeRequestError,
+			},
+		)
+		return
+	}
+
+	esdtData, err := facade.GetESDTData(addr, tokenIdentifier, nonceAsBigInt.Uint64())
+	if err != nil {
+		c.JSON(
+			http.StatusInternalServerError,
+			shared.GenericAPIResponse{
+				Data:  nil,
+				Error: fmt.Sprintf("%s: %s", errors.ErrGetESDTBalance.Error(), err.Error()),
+				Code:  shared.ReturnCodeInternalError,
+			},
+		)
+		return
+	}
+
+	tokenData := esdtNFTTokenData{
+		TokenIdentifier: tokenIdentifier,
+		Balance:         esdtData.Value.String(),
+		Properties:      string(esdtData.Properties),
+	}
+	if esdtData.TokenMetaData != nil {
+		tokenData.Name = string(esdtData.TokenMetaData.Name)
+		tokenData.Creator = string(esdtData.TokenMetaData.Creator)
```

Comment:
this would result in unreadable characters. We should return a bech32 encoded address here. So I would suggest doing this parse inside the node and return a different type from node/facade and return it here directly

Reply:
added it by changing the internal structure in esdt data.